### ISSUE TITLE
Differences between loading source vs loading raw bytecode

### ISSUE THREAD

**Issue by Lyrth** [LuaJIT/LuaJIT#1155]
Hi, first time posting here so please pardon me if I've made some mistakes! Unsure if this has been mentioned already before too.

It seems that loading and running raw bytecode generated from `luajit -b` can have a large performance impact (JIT struggling/turning off?) compared to just running the source directly, given specific code below (other Lua programs I tested this on doesn't seem to show this behavior).

For this specific Lua source: (`cat test.lua`) (Yes, I understand this isn't really efficient code with the table creations, it's only here for demonstration -- I still shouldn't expect bytecode to be significantly 'slower', should I?)
```lua
print"1"

local a = { b = 0 }

while a.b ~= 5e7 do
    a = { b = a.b + 1 }
end

print(a.b)
print"2"
```

On an x86_64 (Windows) machine and freshly-built LuaJIT (MSVC, VS2022), I get a big slowdown running bytecode, as if without JIT:
```bash
$ uname -mo
x86_64 Msys

$ luajit -v
LuaJIT 2.1.1706185428 -- Copyright (C) 2005-2023 Mike Pall. https://luajit.org/

$ luajit -b test.lua test.ljbc

$ time luajit test.lua    # Source
1
50000000
2

real    0m0.144s
user    0m0.000s
sys     0m0.015s

$ time luajit test.ljbc    # Bytecode
1
50000000
2

real    0m3.559s
user    0m0.000s
sys     0m0.015s
```

To double-check, I also ran the same test on an AArch64 system (with the older v2.1.0-beta3), and the same can be observed:
```bash 
$ uname -mo
aarch64 GNU/Linux

$ luajit -v
LuaJIT 2.1.0-beta3 -- Copyright (C) 2005-2022 Mike Pall. https://luajit.org/

$ luajit -b test.lua test.ljbc

$ time luajit test.lua    # Source
1
50000000
2

real    0m0.062s
user    0m0.060s
sys     0m0.000s

$ time luajit test.ljbc    # Bytecode
1
50000000
2

real    0m9.314s
user    0m9.304s
sys     0m0.000s
```

For generating bytecode, I've used `-b` along with `s`, `g`, `W`, `X`, or `d`, but none ran similarly to source still.

**Comment by corsix** [comment]
I hadn't noticed it before, but it looks like round-tripping via bytecode will drop non-constant entries out of template tables. Potentially relevant patch:
```diff
diff --git a/src/lj_bcwrite.c b/src/lj_bcwrite.c
index c062dc498b..ddfa46c56d 100644
--- a/src/lj_bcwrite.c
+++ b/src/lj_bcwrite.c
@@ -133,7 +133,7 @@ static void bcwrite_ktab_sorted_hash(BCWriteCtx *ctx, Node *node, MSize nhash)
   TValue **heap = ctx->heap;
   MSize i = nhash;
   for (;; node--) {  /* Build heap. */
-    if (!tvisnil(&node->val)) {
+    if (!tvisnil(&node->key)) {
       bcwrite_ktabk_heap_insert(heap, --i, nhash, &node->key);
       if (i == 0) break;
     }
@@ -163,7 +163,7 @@ static void bcwrite_ktab(BCWriteCtx *ctx, char *p, const GCtab *t)
     MSize i, hmask = t->hmask;
     Node *node = noderef(t->node);
     for (i = 0; i <= hmask; i++)
-      nhash += !tvisnil(&node[i].val);
+      nhash += !tvisnil(&node[i].key);
   }
   /* Write number of array slots and hash slots. */
   p = lj_strfmt_wuleb128(p, narray);
@@ -184,7 +184,7 @@ static void bcwrite_ktab(BCWriteCtx *ctx, char *p, const GCtab *t)
     } else {
       MSize i = nhash;
       for (;; node--)
-	if (!tvisnil(&node->val)) {
+	if (!tvisnil(&node->key)) {
 	  bcwrite_ktabk(ctx, &node->key, 0);
 	  bcwrite_ktabk(ctx, &node->val, 1);
 	  if (--i == 0) break;
```
As background, the `{ b = a.b + 1 }` in your source becomes a template table `{b = nil}` and then gets the proper value for `b` filled in later (by evaluating `a.b + 1`). Going to bytecode currently reduces that template table to `{}`, at which point filling in `b` requires both adding a key and evaluating the value, rather than evaluating the value into a pre-existing key.

The smaller template table is non-ideal, but _shouldn't_ by itself cause a catastrophic performance degradation. If `-DLUAJIT_ENABLE_TABLE_BUMP` is specified at build time (along with the fixes for it from #606), then the template table gets restored to its original state fairly quickly, and performance stays good.

Table bump is a nice-to-have, but again _shouldn't_ be necessary to restore performance in this case. It looks like the next culprit in the chain is the allocation sinking optimisation; the original source code has a table object being created every loop iteration, and the allocation sinking optimisation is able to completely eliminate that, leading to an allocation-free loop body. The version going via bytecode has "add a key" as an extra step in the loop body, which seems in this case to defeat allocation sinking, so the loop body continues to have an allocation in it, and therefore also requires a GC check in it. Creating and subsequently cleaning up all this garbage is part of why the performance tanks.

For reference, the source code version:
```
---- TRACE 1 IR
....              SNAP   #0   [ ---- ---- ---- ]
0001  {sink}+ tab TDUP   {0x0102cc1090}
0002 x4    >  tab SLOAD  #2    T
0003 x5       int FLOAD  0002  tab.hmask
0004       >  int EQ     0003  +1  
0005 x1       p64 FLOAD  0002  tab.node
0006       >  p64 HREFK  0005  "b"  @1
0007 x28   >  int HLOAD  0006
0008 x28   >+ int ADDOV  0007  +1  
0009          p64 FLOAD  0001  tab.node
0010          p64 HREFK  0009  "b"  @1
0011  {0001}  int HSTORE 0010  0008
....              SNAP   #1   [ ---- ---- 0001 ]
0012       >  int NE     0008  +50000000
....              SNAP   #2   [ ---- ---- 0001 ]
0013 ------------ LOOP ------------
0014  {sink}+ tab TDUP   {0x0102cc1090}
0015 x28   >+ int ADDOV  0008  +1  
0016          p64 FLOAD  0014  tab.node
0017          p64 HREFK  0016  "b"  @1
0018  {0014}  int HSTORE 0017  0015
....              SNAP   #3   [ ---- ---- 0014 ]
0019       >  int NE     0015  +50000000
0020  {sink}  tab PHI    0001  0014
0021 x28      int PHI    0008  0015
0022 x27      nil RENAME 0008  #2  
---- TRACE 1 mcode 132
102ce7f60  mov   x3, #25248
102ce7f64  movk  x3, #714, lsl #16
102ce7f68  movk  x3, #32769, lsl #32
102ce7f6c  movk  x3, #65533, lsl #48
102ce7f70  mov   w2, #65529, lsl #16
102ce7f74  adrp  x0, 0x02be823c
102ce7f78  add   x0, x0, #128
102ce7f7c  ldr   x4, [x19]
102ce7f80  asr   x30, x4, #47
102ce7f84  cmn   x30, #12
102ce7f88  bne   0x02ce7ff0	->0
102ce7f8c  and   x4, x4, #0x7fffffffffff
102ce7f90  ldr   w5, [x4, #52]
102ce7f94  cmp   w5, #1
102ce7f98  bne   0x02ce7ff0	->0
102ce7f9c  ldr   x1, [x4, #40]
102ce7fa0  ldr   x30, [x1, #32]
102ce7fa4  cmp   x30, x3
102ce7fa8  bne   0x02ce7ff0	->0
102ce7fac  ldr   x28, [x1, #24]
102ce7fb0  cmp   x2, x28, lsr #32
102ce7fb4  bne   0x02ce7ff0	->0
102ce7fb8  mov   w28, w28
102ce7fbc  adds  w28, w28, #1
102ce7fc0  bvs   0x02ce7ff0	->0
102ce7fc4  cmp   w28, w0
102ce7fc8  beq   0x02ce7ff4	->1
->LOOP:
102ce7fcc  mov   x27, x28
102ce7fd0  adds  w28, w27, #1
102ce7fd4  bvs   0x02ce7ff8	->2
102ce7fd8  cmp   w28, w0
102ce7fdc  bne   0x02ce7fcc	->LOOP
102ce7fe0  b     0x02ce7ffc	->3
---- TRACE 1 stop -> loop
```
Compared to the bytecode version:
```
---- TRACE 1 IR
....              SNAP   #0   [ ---- ---- ---- ]
0001 [10]  >+ tab TDUP   {0x0104fa9198}
0002 x2    >  tab SLOAD  #2    T
0003 x3       int FLOAD  0002  tab.hmask
0004       >  int EQ     0003  +1  
0005 x24      p64 FLOAD  0002  tab.node
0006       >  p64 HREFK  0005  "b"  @1
0007 x28   >  int HLOAD  0006
0008 x28   >+ int ADDOV  0007  +1  
0009 x23      p64 NEWREF 0001  "b" 
0010          int HSTORE 0009  0008
....              SNAP   #1   [ ---- ---- 0001 ]
0011       >  int NE     0008  +50000000
....              SNAP   #2   [ ---- ---- 0001 ]
0012 ------------ LOOP ------------
0013 [8]   >+ tab TDUP   {0x0104fa9198}
0014 x0       int FLOAD  0001  tab.hmask
0015       >  int EQ     0014  +1  
0016          p64 FLOAD  0001  tab.node
0017 x28   >+ int ADDOV  0008  +1  
0018 x27      p64 NEWREF 0013  "b" 
0019          int HSTORE 0018  0017
....              SNAP   #3   [ ---- ---- 0013 ]
0020       >  int NE     0017  +50000000
0021 x0       tab PHI    0001  0013
0022 x28      int PHI    0008  0017
0023 x26      nil RENAME 0001  #2  
0024 x0       nil RENAME 0001  #2  
---- TRACE 1 mcode 320
104fcfea4  sub   sp, sp, #16
104fcfea8  mov   w25, #65529, lsl #16
104fcfeac  mov   x21, #56888
104fcfeb0  movk  x21, #1272, lsl #16
104fcfeb4  movk  x21, #32769, lsl #32
104fcfeb8  movk  x21, #65533, lsl #48
104fcfebc  mov   x20, #65529, lsl #48
104fcfec0  ldp   x30, x1, [x22, #16]
104fcfec4  cmp   x30, x1
104fcfec8  bls   0x04fcfee0
104fcfecc  mov   w1, #1
104fcfed0  mov   x0, x22
104fcfed4  bl    0x04e6516c	->lj_gc_step_jit
104fcfed8  mov   sp, sp
104fcfedc  cbnz  w0, 0x04fcfff0	->0
104fcfee0  adr   x1, 0x04fa9198
104fcfee4  ldr   x0, [x22, #368]
104fcfee8  bl    0x04e6885c	->lj_tab_dup
104fcfeec  mov   x1, x0
104fcfef0  str   x0, [sp, #16]
104fcfef4  ldr   x0, [x22, #368]
104fcfef8  ldr   x2, [x19]
104fcfefc  asr   x30, x2, #47
104fcff00  cmn   x30, #12
104fcff04  bne   0x04fcfff0	->0
104fcff08  and   x2, x2, #0x7fffffffffff
104fcff0c  mov   w19, #61568
104fcff10  movk  w19, #762, lsl #16
104fcff14  ldr   w3, [x2, #52]
104fcff18  cmp   w3, #1
104fcff1c  bne   0x04fcfff0	->0
104fcff20  ldr   x24, [x2, #40]
104fcff24  ldr   x30, [x24, #32]
104fcff28  cmp   x30, x21
104fcff2c  bne   0x04fcfff0	->0
104fcff30  ldr   x28, [x24, #24]
104fcff34  cmp   x25, x28, lsr #32
104fcff38  bne   0x04fcfff0	->0
104fcff3c  mov   w28, w28
104fcff40  adds  w28, w28, #1
104fcff44  bvs   0x04fcfff0	->0
104fcff48  add   x2, x22, #232
104fcff4c  str   x21, [x2]
104fcff50  bl    0x04e69848	->lj_tab_newkey
104fcff54  mov   x23, x0
104fcff58  ldr   x0, [sp, #16]
104fcff5c  add   x30, x20, w28, uxtw
104fcff60  str   x30, [x23]
104fcff64  cmp   w28, w19
104fcff68  beq   0x04fcfff4	->1
->LOOP:
104fcff6c  mov   x26, x0
104fcff70  ldp   x30, x1, [x22, #16]
104fcff74  cmp   x30, x1
104fcff78  bls   0x04fcff90
104fcff7c  mov   w1, #1
104fcff80  mov   x0, x22
104fcff84  bl    0x04e6516c	->lj_gc_step_jit
104fcff88  mov   sp, sp
104fcff8c  cbnz  w0, 0x04fcfff8	->2
104fcff90  adr   x1, 0x04fa9198
104fcff94  ldr   x0, [x22, #368]
104fcff98  bl    0x04e6885c	->lj_tab_dup
104fcff9c  mov   x1, x0
104fcffa0  str   x0, [sp, #8]
104fcffa4  ldr   w0, [x26, #52]
104fcffa8  cmp   w0, #1
104fcffac  bne   0x04fcfff8	->2
104fcffb0  ldr   x0, [x22, #368]
104fcffb4  adds  w28, w28, #1
104fcffb8  bvs   0x04fcfff8	->2
104fcffbc  add   x2, x22, #232
104fcffc0  str   x21, [x2]
104fcffc4  bl    0x04e69848	->lj_tab_newkey
104fcffc8  mov   x27, x0
104fcffcc  ldr   x0, [sp, #8]
104fcffd0  add   x30, x20, w28, uxtw
104fcffd4  str   x30, [x27]
104fcffd8  cmp   w28, w19
104fcffdc  bne   0x04fcff6c	->LOOP
104fcffe0  b     0x04fcfffc	->3
---- TRACE 1 stop -> loop
```

**Comment by MikePall** [comment]
Fixed. Thanks!
